File: scontrini-frontend/types/receipt.ts
// types/receipt.ts
/**
 * Types per gestione scontrini con prodotti normalizzati
 */

// ===================================
// STORE TYPES
// ===================================

export interface StoreData {
  id: string
  name: string
  chain?: string
  branch_name?: string
  vat_number?: string
  company_name?: string
  address_full?: string
  address_city?: string
  address_province?: string
  is_mock: boolean
}

// ===================================
// NORMALIZED PRODUCT TYPES
// ===================================

export interface NormalizedProduct {
  id: string
  canonical_name: string
  brand: string | null
  category: string
  subcategory: string | null
  size: string
  unit_type: string
  confidence: number
  pending_review: boolean
  from_cache?: boolean
}

// ===================================
// RECEIPT ITEM TYPES
// ===================================

export interface ReceiptItem {
  raw_product_name: string
  quantity: number
  unit_price: number
  total_price: number
  category?: string
}

export interface ReceiptItemWithNormalized extends ReceiptItem {
  // ID per tracking
  id?: string
  receipt_item_id?: string
  
  // Dati normalizzati
  normalized_product_id?: string
  canonical_name?: string
  brand?: string | null
  subcategory?: string | null
  size?: string
  unit_type?: string
  confidence?: number
  pending_review?: boolean
  from_cache?: boolean
  user_verified?: boolean
}

// ===================================
// PARSED RECEIPT
// ===================================

export interface ParsedReceipt {
  // Store info
  store_id?: string
  store_name?: string
  company_name?: string
  vat_number?: string
  store_address?: string
  store_data?: StoreData
  
  // Receipt info
  receipt_date?: string
  receipt_time?: string
  total_amount?: number
  tax_amount?: number
  payment_method?: string
  discount_amount?: number
  
  // Items con normalizzazione
  items: ReceiptItemWithNormalized[]
}

// ===================================
// API REQUEST/RESPONSE
// ===================================

export interface ProcessReceiptRequest {
  household_id: string
  uploaded_by: string
  image_url: string
}

export interface ProcessReceiptResponse {
  success: boolean
  receipt_id?: string
  parsed_data?: ParsedReceipt
  ocr_confidence?: number
  error?: string
  message?: string
}

export interface UpdateProductReviewRequest {
  canonical_name: string
  brand?: string | null
  category: string
  subcategory?: string | null
  size?: string
  unit_type?: string
}

export interface UpdateProductReviewResponse {
  success: boolean
  message?: string
  normalized_product_id?: string
}

// ===================================
// RECEIPT ENTITY
// ===================================

export interface Receipt {
  id: string
  household_id: string
  uploaded_by: string
  image_url: string
  store_id?: string
  store_name?: string
  store_address?: string
  receipt_date?: string
  receipt_time?: string
  total_amount?: number
  tax_amount?: number
  payment_method?: string
  processing_status: 'pending' | 'processing' | 'completed' | 'failed'
  ocr_confidence?: number
  created_at: string
  updated_at: string
}

// ===================================
// UI HELPER TYPES
// ===================================

export type UploadStep = 'upload' | 'processing' | 'review' | 'complete'

export interface ProductEditState {
  receipt_item_id: string
  canonical_name: string
  brand: string
  category: string
  subcategory: string
  size: string
  unit_type: string
}

File: scontrini-backend/app/api/schemas/receipt.py
"""
Aggiorna app/api/schemas/receipt.py con questi schemas
"""
from pydantic import BaseModel
from typing import List, Optional
from datetime import date, time

# ===================================
# REQUEST SCHEMAS
# ===================================

class ProcessReceiptRequest(BaseModel):
    """Request per processare scontrino"""
    household_id: str
    uploaded_by: str
    image_url: str


# ===================================
# RESPONSE SCHEMAS
# ===================================

class ReceiptItemData(BaseModel):
    """Dati singolo item con normalizzazione"""
    # Dati grezzi
    raw_product_name: str
    quantity: float
    unit_price: float
    total_price: float
    
    # Dati normalizzati
    normalized_product_id: Optional[str] = None
    canonical_name: Optional[str] = None
    brand: Optional[str] = None
    category: Optional[str] = None
    subcategory: Optional[str] = None
    size: Optional[str] = None
    unit_type: Optional[str] = None
    confidence: Optional[float] = None
    pending_review: Optional[bool] = False
    from_cache: Optional[bool] = False


class StoreData(BaseModel):
    """Dati store completi"""
    id: str
    name: str
    chain: Optional[str] = None
    branch_name: Optional[str] = None
    vat_number: Optional[str] = None
    company_name: Optional[str] = None
    address_full: Optional[str] = None
    address_city: Optional[str] = None
    address_province: Optional[str] = None
    is_mock: bool = False


class ParsedReceiptData(BaseModel):
    """Dati scontrino parsato con store e normalizzazione"""
    # Store info
    store_id: Optional[str] = None
    store_name: Optional[str] = None
    company_name: Optional[str] = None
    vat_number: Optional[str] = None
    store_address: Optional[str] = None
    store_data: Optional[dict] = None  # Oggetto store completo
    
    # Receipt info
    receipt_date: Optional[str] = None
    receipt_time: Optional[str] = None
    total_amount: Optional[float] = None
    payment_method: Optional[str] = None
    discount_amount: Optional[float] = None
    
    # Items con normalizzazione
    items: List[ReceiptItemData] = []


class ProcessReceiptResponse(BaseModel):
    """Response processamento scontrino"""
    success: bool
    receipt_id: Optional[str] = None
    parsed_data: Optional[ParsedReceiptData] = None
    ocr_confidence: Optional[float] = None
    error: Optional[str] = None
    message: Optional[str] = None


# ===================================
# UPDATE REVIEW SCHEMAS
# ===================================

class UpdateProductReviewRequest(BaseModel):
    """Request per aggiornare prodotto dopo review utente"""
    # Dati corretti dall'utente
    canonical_name: str
    brand: Optional[str] = None
    category: str
    subcategory: Optional[str] = None
    size: Optional[str] = None
    unit_type: Optional[str] = None


class UpdateProductReviewResponse(BaseModel):
    """Response update review"""
    success: bool
    message: Optional[str] = None
    normalized_product_id: Optional[str] = None


# ===================================
# ALTRI SCHEMAS (già esistenti, mantenere)
# ===================================

class ReceiptResponse(BaseModel):
    """Response singolo receipt"""
    id: str
    household_id: str
    uploaded_by: str
    image_url: str
    store_id: Optional[str] = None
    store_name: Optional[str] = None
    store_address: Optional[str] = None
    receipt_date: Optional[str] = None
    receipt_time: Optional[str] = None
    total_amount: Optional[float] = None
    payment_method: Optional[str] = None
    processing_status: str
    ocr_confidence: Optional[float] = None
    created_at: str
    updated_at: str
    items: List[dict] = []


class ReceiptListResponse(BaseModel):
    """Response lista receipts"""
    receipts: List[ReceiptResponse]
    total: int
    page: int
    page_size: int